File: client/app/src/main/java/com/example/client/model/data/Message.kt
// Message.kt
package com.example.client.model.data
import org.json.JSONObject

data class Message(
    val id: String = "",
    val roomId: String = "",
    val senderId: String = "",
    val content: String = "",
    val type: String = "TEXT",     // "TEXT", "IMAGE", "FILE"
    val fileName: String? = null,  // <--- 1. THÊM DÒNG NÀY (để lưu tên file PDF/Word)
    val createdAt: String = "",
    val timestamp: Long = 0L,
    val status: String = "sent"
) {
    companion object {
        fun fromJson(json: JSONObject): Message {
            return Message(
                // Lưu ý: MongoDB thường trả về "_id", kiểm tra lại server gửi "id" hay "_id"
                id = if (json.has("id")) json.optString("id") else json.optString("_id"),

                roomId = json.optString("roomId"),
                senderId = if (json.has("senderId")) json.optString("senderId") else json.optString("sender"),
                content = json.optString("content"),
                type = json.optString("type", "TEXT"),

                // <--- 2. THÊM DÒNG NÀY ĐỂ LẤY TÊN FILE TỪ JSON
                fileName = json.optString("fileName", null),

                createdAt = json.optString("createdAt", ""),
                timestamp = json.optLong("timestamp", System.currentTimeMillis()),
                status = json.optString("status", "sent")
            )
        }
    }
}

File: server/src/models/Message.js
// server/src/models/Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
    roomId: {
        type: String, 
        required: true
    },
    senderId: {
        type: String, 
        required: true
    },
    content: {
        type: String,
        required: true
    },
    type: {
        type: String,
        // Enum phải viết HOA để khớp với Client gửi lên (TEXT, IMAGE, FILE)
        enum: ['TEXT', 'IMAGE', 'FILE', 'VOICE', 'CONTACT', 'SYSTEM'], 
        default: 'TEXT'
    },
    // 👇 QUAN TRỌNG: Thêm trường này để lưu tên file (PDF/Word)
    fileName: { 
        type: String, 
        default: null 
    },
    status: {
        type: String,
        enum: ['sending', 'sent', 'delivered', 'seen'],
        default: 'sent'
    },
    timestamp: { 
        type: Number, 
        default: () => Date.now() 
    }
}, { 
    timestamps: true 
});

// Helper format dữ liệu trả về cho Android
messageSchema.methods.formatForClient = function() {
    const date = new Date(this.timestamp || this.createdAt);
    const timeString = date.getHours().toString().padStart(2, '0') + ':' + 
                       date.getMinutes().toString().padStart(2, '0');

    return {
        id: this._id ? this._id.toString() : Date.now().toString(),
        
        // 👇 SỬA LẠI CHO KHỚP SCHEMA (quan trọng)
        roomId: this.roomId,      // Schema là roomId -> dùng this.roomId
        senderId: this.senderId,  // Schema là senderId -> dùng this.senderId
        
        content: this.content || "",
        type: this.type || "TEXT",
        
        // Trả về tên file cho Client hiển thị
        fileName: this.fileName || null,
        
        createdAt: timeString,
        timestamp: this.timestamp,
        status: this.status
    };
};

const Message = mongoose.model('Message', messageSchema);
module.exports = Message;